src/zarr/core/array.py

```diff
@@ -1031,7 +1049,10 @@ def dtype(self) -> np.dtype[Any]:
         np.dtype
             Data type of the array
         """
-        return self.metadata.dtype
+        if self.metadata.zarr_format == 2:
+            return self.metadata.dtype.to_dtype()
+        else:
+            return self.metadata.data_type.to_dtype()
```

Comment:
I thought we were tryiing to keep the `Metadata.dtype` interface common between zarr_format 2 and 3? Why this change?

Reply:
First, this change made some of the other changes in this PR easier (I can't remember what became easier exactly) And from a principled POV, I don't think the two metadata documents should have linked APIs. They are logically independent entities, and it should be the job of the routines that consume the metadata documents to generate a abstracted representation that hides the v2 / v3 differences. The `AsyncArray` class is perfect for this.